## Import & Install packges

In [1]:
!pip3 install nltk

In [2]:
!pip3 install --upgrade pip

In [3]:
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
import numpy as np
import pandas as pd
import re
from sklearn.feature_extraction.text import CountVectorizer , TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.pipeline import make_pipeline
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV

nltk.download('stopwords')
stop_words = set(stopwords.words('english'))
nltk.download('punkt')

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/ahmedbanafa/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     /Users/ahmedbanafa/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [4]:
nltk.download()

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

## Read the data

In [4]:
df = pd.read_csv('train.csv')

In [5]:
df

id keyword location  \
0         1     NaN      NaN   
1         4     NaN      NaN   
2         5     NaN      NaN   
3         6     NaN      NaN   
4         7     NaN      NaN   
...     ...     ...      ...   
7608  10869     NaN      NaN   
7609  10870     NaN      NaN   
7610  10871     NaN      NaN   
7611  10872     NaN      NaN   
7612  10873     NaN      NaN   

                                                   text  target  
0     Our Deeds are the Reason of this #earthquake M...       1  
1                Forest fire near La Ronge Sask. Canada       1  
2     All residents asked to 'shelter in place' are ...       1  
3     13,000 people receive #wildfires evacuation or...       1  
4     Just got sent this photo from Ruby #Alaska as ...       1  
...                                                 ...     ...  
7608  Two giant cranes holding a bridge collapse int...       1  
7609  @aria_ahrary @TheTawniest The out of control w...       1  
7610  M1.94 [01:04 UTC]?5km S of Volcano Hawaii. htt...       1  
7611  Police investigating after an e-bike collided ...       1  
7612  The Latest: More Homes Razed by Northern Calif...       1  

[7613 rows x 5 columns]

## Cleaning & preprocess data

In [6]:
# Assuming 'df' is your DataFrame containing the tweets
text_data = df['text'].values

corpus = []
ps = PorterStemmer()
all_stopwords = stopwords.words('english')
all_stopwords.remove('not')

for i in range(len(text_data)):
    text = text_data[i]
    
    # Ensure the text is a string (decode if necessary)
    if isinstance(text, bytes):
        text = text.decode('utf-8', errors='ignore')
    
    # Remove URLs
    text = re.sub(r'http\S+|www\S+|https\S+', '', text)
    
    # Remove HTML tags
    text = re.sub(r'<.*?>', '', text)
    
    # Remove special characters and numbers
    text = re.sub(r'[^A-Za-z\s]', '', text)
    
    # Convert to lowercase
    text = text.lower()
    
    # Tokenize text
    text = nltk.word_tokenize(text)
    
    # Stem words and remove stopwords
    text = [ps.stem(word) for word in text if word not in all_stopwords]
    
    # Join the cleaned words back into a single string
    cleaned_text = ' '.join(text)
    
    # Add the cleaned text to the corpus
    corpus.append(cleaned_text)

print(corpus)

['deed reason earthquak may allah forgiv us', 'forest fire near la rong sask canada', 'resid ask shelter place notifi offic evacu shelter place order expect', 'peopl receiv wildfir evacu order california', 'got sent photo rubi alaska smoke wildfir pour school', 'rockyfir updat california hwi close direct due lake counti fire cafir wildfir', 'flood disast heavi rain caus flash flood street manit colorado spring area', 'im top hill see fire wood', 'there emerg evacu happen build across street', 'im afraid tornado come area', 'three peopl die heat wave far', 'haha south tampa get flood hah wait second live south tampa gon na gon na fvck flood', 'rain flood florida tampabay tampa day ive lost count', 'flood bago myanmar arriv bago', 'damag school bu multi car crash break', 'what man', 'love fruit', 'summer love', 'car fast', 'goooooooaaaaaal', 'ridicul', 'london cool', 'love ski', 'wonder day', 'looooool', 'wayi cant eat shit', 'nyc last week', 'love girlfriend', 'cooool', 'like pasta', 'e

In [7]:
corpus

['deed reason earthquak may allah forgiv us',
 'forest fire near la rong sask canada',
 'resid ask shelter place notifi offic evacu shelter place order expect',
 'peopl receiv wildfir evacu order california',
 'got sent photo rubi alaska smoke wildfir pour school',
 'rockyfir updat california hwi close direct due lake counti fire cafir wildfir',
 'flood disast heavi rain caus flash flood street manit colorado spring area',
 'im top hill see fire wood',
 'there emerg evacu happen build across street',
 'im afraid tornado come area',
 'three peopl die heat wave far',
 'haha south tampa get flood hah wait second live south tampa gon na gon na fvck flood',
 'rain flood florida tampabay tampa day ive lost count',
 'flood bago myanmar arriv bago',
 'damag school bu multi car crash break',
 'what man',
 'love fruit',
 'summer love',
 'car fast',
 'goooooooaaaaaal',
 'ridicul',
 'london cool',
 'love ski',
 'wonder day',
 'looooool',
 'wayi cant eat shit',
 'nyc last week',
 'love girlfriend',

In [8]:
cv = CountVectorizer()
dfcv = cv.fit_transform(corpus)
print(dfcv)

  (0, 3010)	1
  (0, 9778)	1
  (0, 3594)	1
  (0, 7435)	1
  (0, 325)	1
  (0, 4438)	1
  (0, 12754)	1
  (1, 4431)	1
  (1, 4282)	1
  (1, 8147)	1
  (1, 6679)	1
  (1, 10193)	1
  (1, 10440)	1
  (1, 1797)	1
  (2, 9965)	1
  (2, 694)	1
  (2, 10754)	2
  (2, 9161)	2
  (2, 8395)	1
  (2, 8527)	1
  (2, 3939)	1
  (2, 8693)	1
  (2, 4003)	1
  (3, 3939)	1
  (3, 8693)	1
  :	:
  (7609, 12044)	1
  (7610, 12977)	1
  (7610, 5246)	1
  (7610, 12786)	1
  (7611, 1844)	1
  (7611, 9240)	1
  (7611, 10062)	1
  (7611, 5912)	1
  (7611, 6995)	1
  (7611, 10656)	1
  (7611, 6001)	1
  (7611, 12102)	1
  (7611, 11570)	1
  (7611, 2361)	1
  (7611, 3615)	2
  (7611, 9285)	1
  (7611, 8341)	1
  (7612, 13272)	1
  (7612, 1760)	1
  (7612, 5489)	1
  (7612, 8213)	1
  (7612, 6752)	1
  (7612, 8369)	1
  (7612, 23)	1
  (7612, 9734)	1


In [13]:
# x = corpus
# tf = TfidfVectorizer()
# x = tf.fit_transform(corpus)
# print(x)

In [9]:
y = df['target']

In [10]:
df['keyword'].value_counts()

keyword
fatalities               45
deluge                   42
armageddon               42
sinking                  41
damage                   41
                         ..
forest%20fire            19
epicentre                12
threat                   11
inundation               10
radiation%20emergency     9
Name: count, Length: 221, dtype: int64

## Training & Selection model

In [11]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.pipeline import make_pipeline
from sklearn.metrics import classification_report
from sklearn.model_selection import GridSearchCV, train_test_split


# Vectorizer
tf = TfidfVectorizer()
x = tf.fit_transform(corpus)

# Train-test split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=26)

# Models
models = {
    "Naive Bayes": MultinomialNB(),
    "Logistic Regression": LogisticRegression(max_iter=1000),
    "SVM": SVC(kernel='linear'),
    "Random Forest": RandomForestClassifier(),
    "Neural Network": MLPClassifier(max_iter=1000)
    
}

# Train and evaluate each model
for model_name, model in models.items():
    model.fit(x_train, y_train)
    y_pred = model.predict(x_test)
    print(f"Model: {model_name}")
    print(classification_report(y_test, y_pred))

# Hyperparameter grids
param_grids = {
    "Naive Bayes": {
        'alpha': [0.1, 0.5, 1.0]
    },
    "Logistic Regression": {
        'C': [0.1, 1, 10]
    },
    "SVM": {
        'C': [0.1, 1, 10]
    },
    "Random Forest": {
        'n_estimators': [100, 200, 300],
        'max_features': ['auto', 'sqrt', 'log2'],
        'max_depth': [4, 6, 8],
        'criterion': ['gini', 'entropy']
    },
    "Neural Network": {
        'hidden_layer_sizes': [(50,), (100,), (50, 50)],
        'alpha': [0.0001, 0.001, 0.01]
    }
}

# Cross-validation and hyperparameter tuning
best_models = {}
for model_name, model in models.items():
    param_grid = param_grids[model_name]
    grid_search = GridSearchCV(model, param_grid, cv=5, n_jobs=-1)
    grid_search.fit(x_train, y_train)
    best_models[model_name] = grid_search.best_estimator_
    print(f"Best parameters for {model_name}: {grid_search.best_params_}")
    y_pred = grid_search.predict(x_test)
    print(f"Classification report for {model_name}:")
    print(classification_report(y_test, y_pred))


Model: Naive Bayes
              precision    recall  f1-score   support

           0       0.79      0.89      0.84       893
           1       0.82      0.66      0.73       630

    accuracy                           0.80      1523
   macro avg       0.80      0.78      0.78      1523
weighted avg       0.80      0.80      0.79      1523

Model: Logistic Regression
              precision    recall  f1-score   support

           0       0.78      0.91      0.84       893
           1       0.83      0.64      0.72       630

    accuracy                           0.80      1523
   macro avg       0.81      0.77      0.78      1523
weighted avg       0.80      0.80      0.79      1523

Model: SVM
              precision    recall  f1-score   support

           0       0.80      0.89      0.84       893
           1       0.81      0.68      0.74       630

    accuracy                           0.80      1523
   macro avg       0.80      0.78      0.79      1523
weighted avg     

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:540: FitFailedWarning: 
90 fits failed out of a total of 270.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
36 fits failed with the following error:
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/model_selection/_validation.py", line 888, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/base.py", line 1466, in wrapper
    estimator._validate_params()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/

Best parameters for Random Forest: {'criterion': 'gini', 'max_depth': 8, 'max_features': 'sqrt', 'n_estimators': 300}
Classification report for Random Forest:
              precision    recall  f1-score   support

           0       0.62      1.00      0.77       893
           1       1.00      0.15      0.26       630

    accuracy                           0.65      1523
   macro avg       0.81      0.57      0.51      1523
weighted avg       0.78      0.65      0.56      1523

Best parameters for Neural Network: {'alpha': 0.01, 'hidden_layer_sizes': (100,)}
Classification report for Neural Network:
              precision    recall  f1-score   support

           0       0.79      0.84      0.82       893
           1       0.75      0.69      0.72       630

    accuracy                           0.78      1523
   macro avg       0.77      0.76      0.77      1523
weighted avg       0.78      0.78      0.78      1523



Based on the evaluation metrics, the best models are
 #### Naive Bayes, Logistic Regression, and SVM, each with an accuracy of 80%.
  To choose the best model, we should consider additional metrics like precision, recall, and f1-score:

Naive Bayes:
Precision: 0.80
Recall: 0.78
F1-score: 0.79


Logistic Regression:
Precision: 0.81
Recall: 0.77
F1-score: 0.78


SVM:
Precision: 0.80
Recall: 0.78
F1-score: 0.79


The differences in these metrics are minimal, so any of these models would be a good choice.
#### Given Logistic Regression's simplicity and efficiency, we will select it as the best model.

#### Application in Disaster Response and Emergency Management

##### Predictive Analysis: 
Early Warning Systems: Predict future events using historical data.

Resource Allocation: Identify high-risk areas and allocate resources.

##### Post-Disaster Analysis:
Damage Assessment: Analyze social media to assess damage.

Recovery Planning: Plan recovery efforts for the most affected regions.